test/gtest/device_api/common/device_kernels.cu

```diff
@@ -0,0 +1,237 @@
+                                                params.write.channelId,
+                                                params.withNoDelay,
+                                                req_ptr);
+        break;
+
+    case nixl_device_operation_t::SIGNAL_POST:
+        status = nixlGpuPostSignalXferReq<level>(params.reqHandle,
+                                                 params.signalPost.signalDescIndex,
+                                                 params.signalPost.signalInc,
+                                                 params.signalPost.signalOffset,
+                                                 params.signalPost.channelId,
+                                                 params.withNoDelay,
+                                                 req_ptr);
+        break;
+
+    case nixl_device_operation_t::SIGNAL_WAIT: {
+        if (params.signalWait.signalAddr == nullptr) {
+            return NIXL_ERR_INVALID_PARAM;
+        }
+
+        uint64_t value;
+        do {
+            value = nixlGpuReadSignal<level>(params.signalWait.signalAddr);
+        } while (value != params.signalWait.expectedValue);
+
+        return NIXL_SUCCESS;
+    }
+
+    case nixl_device_operation_t::SIGNAL_WRITE:
+        if (params.signalWrite.signalAddr == nullptr) {
+            return NIXL_ERR_INVALID_PARAM;
+        }
+        nixlGpuWriteSignal<level>(params.signalWrite.signalAddr, params.signalWrite.value);
+        return NIXL_SUCCESS;
+
+    default:
+        return NIXL_ERR_INVALID_PARAM;
+    }
+
+    if (status != NIXL_IN_PROG) {
+        return status;
+    }
+
+    if (!params.withNoDelay || (req_ptr == nullptr)) {
+        return NIXL_SUCCESS;
+    }
+
+    do {
+        status = nixlGpuGetXferStatus<level>(*req_ptr);
+    } while (status == NIXL_IN_PROG);
+
+    return status;
+}
+
+template<nixl_gpu_level_t level>
+__device__ nixl_status_t
+kernelJob(const nixlDeviceKernelParams &params, nixlGpuXferStatusH *shared_reqs) {
+    if (blockDim.x > max_threads_per_block) {
+        return NIXL_ERR_INVALID_PARAM;
+    }
+
+    if (params.numIters == 0) {
+        return NIXL_ERR_INVALID_PARAM;
+    }
+
+    nixlGpuXferStatusH *req_ptr = nullptr;
+    if (params.withRequest) {
+        const size_t req_index = getStatusIndex<level>();
+        req_ptr = &shared_reqs[req_index];
+    }
+
+    nixl_status_t status;
+    for (size_t i = 0; i < params.numIters - 1; i++) {
+        status = doOperation<level>(params, req_ptr);
+        if (status != NIXL_SUCCESS) {
+            return status;
+        }
+    }
+
+    // Last iteration forces completion to ensure all operations are finished
+    nixlDeviceKernelParams params_force_completion = params;
+    params_force_completion.withNoDelay = true;
+    nixlGpuXferStatusH *status_ptr = nullptr;
+    if (params.withRequest) {
+        const size_t req_index = getStatusIndex<level>();
+        status_ptr = &shared_reqs[req_index];
+    }
```

Comment:
The same code is repeated twice in this method. Please add a helper method returning `nixlGpuXferStatusH *` to avoid code duplication.